### PR TITLE
add the first commit

### PR DESCRIPTION
**Pull request by ZLI-afk**

### CODE REVIEW

**Review by AnguseZhang — COMMENTED** [review]
I still do not finsh the review. First leave some comments.

**Comment by Chengqian-Zhang** [comment]
Does it feel a little better to have unit tests?

File: apex/ABACUS_flow.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+            parameters={
+                "run_image_config": {"command": self.abacus_run_command},
+                "task_name": relaxmake.outputs.parameters["task_names"],
+                "backward_list": ["OUT.ABACUS","log"],
+                "backward_dir_name": "relax_task",
+                "log_name": "log"
+            },
+            artifacts={
+                "task_path": relaxmake.outputs.artifacts["task_paths"],
+                "optional_artifact": upload_artifact({"pp_orb": "./"})
```
Comment (zjgemi) [inline apex/ABACUS_flow.py:130]: What is the artifact `pp_orb` for? And is the path `./` hard-coded?

File: apex/calculator/Lammps.py [hunk 2]
```diff
@@ -0,0 +1,532 @@
...
+            with open(self.in_lammps, "r") as fin:
+                fc = fin.read()
+        # user input in.lammps for apex calculation
+        if "input_prop" in cal_setting and os.path.isfile(cal_setting["input_prop"]):
+            with open(os.path.abspath(cal_setting["input_prop"]), "r") as fin:
+                fc = fin.read()
+
+        else:
+            """
+            flowing dlog.info are replaced as print
```
Comment (zjgemi) [inline apex/calculator/Lammps.py:154]: logging is also OK, but not print

File: README.md [hunk 3]
```diff
@@ -1 +1,38 @@
-APEX: Alloy Properties EXplorer using simulations
+# APEX: Alloy Properties EXplorer using simulations
+## General alloy properties test with VASP, ABACUS, and LAMMPS.
+This workflow is a part of [AI Square](https://aissquare.com/). We want to refract the autotest code based on dflow.
```
[the remark this answers is not in the document]
Comment (ZLI-afk) [reply README.md]: revised.

File: apex/LAMMPS_OPs.py [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+        return OPIOSign({
+            'output_lammps': Artifact(Path, sub_path=False)
+        })
+
+    @OP.exec_sign_check
+    def execute(self, op_in: OPIO) -> OPIO:
+        cwd = os.getcwd()
+        os.chdir(op_in["input_lammps"])
+        cmd = op_in["run_command"]
+        subprocess.call(cmd, shell=True)
```
[the remark this answers is not in the document]
Comment (ZLI-afk) [reply apex/LAMMPS_OPs.py]: revised.

File: apex/calculator/lib/abacus_scf.py [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+    return ret
+
+
+def make_abacus_scf_input(fp_params):
+    # Make INPUT file for abacus pw scf calculation.
+    ret = "INPUT_PARAMETERS\n"
+    ret += "calculation scf\n"
+    for key in fp_params:
+        if key == "ecutwfc":
+            fp_params["ecutwfc"] = float(fp_params["ecutwfc"])
```
[the remark this answers is not in the document]
Comment (ZLI-afk) [reply apex/calculator/lib/abacus_scf.py]: revised to dargs.

File: apex/LAMMPS_flow.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        )
+        self.propscal = propscal
+
+        propspost = Step(
+            name="Propspost",
+            template=PythonOPTemplate(PropsPostLAMMPS, image=self.dpgen_image_name, command=["python3"]),
+            artifacts={"input_post": propscal.outputs.artifacts["output_lammps"],
+                       "input_all": propsmake.outputs.artifacts["output"],
+                       "param": upload_artifact(self.props_param)},
+            parameters={"path": cwd}
```
Comment (zjgemi) [inline apex/LAMMPS_flow.py:160]: It's better to set key for each step in case of reusing.

**Comment by zjgemi** [comment]
Consider to add UTs to CI workflow?

File: apex/LAMMPS_flow.py (hunk 6 above)
Comment (kevinwenminion) [reply apex/LAMMPS_flow.py]: Keys were added.

Comment (ZLI-afk) [reply apex/LAMMPS_flow.py]: revised

**Comment by kevinwenminion** [comment]
> Consider to add UTs to CI workflow?

added

File: apex/calculator/ABACUS.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+
+        # user input INCAR for apex calculation
+        if "input_prop" in cal_setting and os.path.isfile(cal_setting["input_prop"]):
+            incar_prop = os.path.abspath(cal_setting["input_prop"])
+            incar = abacus_scf.get_abacus_input_parameters(incar_prop)
+            """dlog.info(
+                "Detected 'input_prop' in 'relaxation', use %s as INPUT, and ignore 'cal_setting'"
+                % incar_prop
+            )"""
+            print(
```
[the remark this answers is not in the document]
Comment (ZLI-afk) [reply apex/calculator/ABACUS.py]: revised. forgot to change it back to logging, as APEX needs to be independent of the previous degen.

**Comment by ZLI-afk** [comment]
> Does it feel a little better to have unit tests?

added

**Comment by Chengqian-Zhang** [comment]
I don't find the integration test which may need dflow debug mode. Is it necessary to do it?

File: apex/__main__.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
+#!/usr/bin/env python
+
+from dflow import config, s3_config
+from dflow.plugins import bohrium
+from dflow.plugins.bohrium import TiefblueClient
+from monty.serialization import loadfn
+
+config["host"] = "https://workflows.deepmodeling.com"
```
[the remark this answers is not in the document]
Comment (ZLI-afk) [reply apex/__main__.py]: Revised settings into the global.json

File: tests/test_lammps.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        self.assertEqual(model_param, self.Lammps.model_param)
+
+    def test_make_potential_files(self):
+        cwd = os.getcwd()
+        abs_equi_path = os.path.abspath(self.equi_path)
+        self.Lammps.make_potential_files(abs_equi_path)
+        self.assertTrue(os.path.islink(os.path.join(self.equi_path, "frozen_model.pb")))
+        self.assertTrue(os.path.isfile(os.path.join(self.equi_path, "inter.json")))
+        ret = loadfn(os.path.join(self.equi_path, "inter.json"))
+        self.assertTrue(self.inter_param, ret)
```
[the remark this answers is not in the document]
Comment (ZLI-afk) [reply tests/test_lammps.py]: revised

File: apex/VASP_flow.py [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+
+        self.run_step_config_relax = {
+            "executor": {
+                "type": "dispatcher",
+                "image_pull_policy": "IfNotPresent",
+                "machine_dict": {
+                    "batch_type": self.batch_type,
+                    "context_type": self.context_type,
+                    "remote_profile": {
+                        "email": self.email,
```
[the remark this answers is not in the document]
Comment (ZLI-afk) [reply apex/VASP_flow.py:72]: revised to be independent on Bohrium setting

File: tests/test_refine.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+                "type": "eos",
+                "skip": False,
+                "vol_start": 0.8,
+                "vol_end": 1.2,
+                "vol_step": 0.01,
+                "cal_setting": {
+                    "relax_pos": True,
+                    "relax_shape": True,
+                    "relax_vol": False,
+                    "overwrite_interaction": {
```
[the remark this answers is not in the document]
Comment (kevinwenminion) [reply tests/test_refine.py:42]: This was considered in the last version of Autotest. The reason for "overwrite_interaction" is that sometimes we want to use other interatomic potentials or DFT to calculate some properties. The content of "overwrite_interaction" part is indeed the same as "interaction" part.

**Pull request merged**